src/audio/copier.c

```diff
@@ -0,0 +1,666 @@
+			i = IPC4_SINK_QUEUE_ID(sink->id);
+			buffer_invalidate(src, in_size);
+			cd->converter[i](&src->stream, 0, &sink->stream, 0, sample);
+			buffer_writeback(sink, out_size);
+
+			comp_update_buffer_produce(sink, out_size);
+		}
+
+		comp_update_buffer_consume(src, in_size);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static void update_internal_comp(struct comp_dev *parent, struct comp_dev *child)
+{
+	child->period = parent->period;
+	child->pipeline = parent->pipeline;
+	child->priority = parent->priority;
+	child->direction = parent->direction;
+}
+
+/* configure the DMA params */
+static int copier_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	struct copier_data *cd = comp_get_drvdata(dev);
+	union ipc4_connector_node_id node_id;
+	struct comp_buffer *buffer;
+	struct comp_buffer *sink;
+	struct list_item *sink_list;
+	int ret, dir;
+
+	comp_err(dev, "copier_params()");
+
+	memset(params, 0, sizeof(*params));
+	params->direction = cd->direction;
+	params->channels = cd->config.base.audio_fmt.channels_count;
+	params->rate = cd->config.base.audio_fmt.sampling_frequency;
+	params->sample_container_bytes = cd->config.base.audio_fmt.depth;
+	params->sample_valid_bytes = cd->config.base.audio_fmt.valid_bit_depth;
+
+	node_id.dw = cd->config.gtw_cfg.node_id;
+	params->stream_tag = node_id.f.v_index + 1;
+	params->frame_fmt = dev->ipc_config.frame_fmt;
+	params->buffer_fmt = cd->config.base.audio_fmt.interleaving_style;
+	params->buffer.size = cd->config.base.ibs;
+
+	list_for_item(sink_list, &dev->bsink_list) {
+		int  i, j;
+
+		sink = container_of(sink_list, struct comp_buffer, source_list);
+		j = IPC4_SINK_QUEUE_ID(sink->id);
+		sink->stream.channels = cd->out_fmt[j].channels_count;
+		sink->stream.rate = cd->out_fmt[j].sampling_frequency;
+		sink->stream.frame_fmt  = (cd->out_fmt[j].valid_bit_depth >> 3) - 2;
+		sink->buffer_fmt = cd->out_fmt[j].interleaving_style;
+
+		for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+			sink->chmap[i] = (cd->out_fmt[j].ch_map >> i * 4) & 0xf;
+	}
+
+	if (!list_is_empty(&dev->bsink_list)) {
+		dir = PPL_CONN_DIR_COMP_TO_BUFFER;
+		buffer = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		list_item_del(buffer->source_list.next);
+		list_item_del(dev->bsink_list.next);
+	} else {
+		dir = PPL_CONN_DIR_BUFFER_TO_COMP;
+		buffer = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+		list_item_del(buffer->sink_list.next);
+		list_item_del(dev->bsource_list.next);
+	}
+
+	if (cd->host) {
+		update_internal_comp(dev, cd->host);
+		connect_comp_to_buffer(cd->host, buffer, dir);
+		ret = cd->host->drv->ops.params(cd->host, params);
+		connect_comp_to_buffer(dev, buffer, dir);
+	} else if (cd->dai) {
+		update_internal_comp(dev, cd->dai);
+		connect_comp_to_buffer(cd->dai, buffer, dir);
+		ret = cd->dai->drv->ops.params(cd->dai, params);
+		connect_comp_to_buffer(dev, buffer, dir);
+	} else {
+		ret = 0;
+	}
```

Comment:
We are seeing this code block in lots of places. Could be simplified with an inline function.